logstreams/src/main/java/io/zeebe/logstreams/impl/log/index/LogBlockIndex.java

```diff
@@ -42,222 +41,163 @@
+            return true;
+          } else {
+            return false;
+          }
+        });
 
-    return idx;
+    return blockPosition.get();
   }
 
-  /**
-   * Invoked by the log Appender thread after it has first written one or more entries to a block.
-   *
-   * @param logPosition the virtual position of the block (equal or smaller to the v position of the
-   *     first entry in the block)
-   * @param storageAddr the physical address of the block in the underlying storage
-   * @return the new size of the index.
-   */
-  public int addBlock(long logPosition, long storageAddr) {
-    final int currentIndexSize =
-        indexBuffer.getInt(indexSizeOffset()); // volatile get not necessary
-    final int entryOffset = entryOffset(currentIndexSize);
-    final int newIndexSize = 1 + currentIndexSize;
-
-    if (newIndexSize > capacity) {
-      expandIndexBuffer();
-    }
-
-    if (lastVirtualPosition >= logPosition) {
+  public synchronized void addBlock(long blockPosition, long blockAddress) {
+    if (lastVirtualPosition >= blockPosition) {
       final String errorMessage =
           String.format(
               "Illegal value for position.Value=%d, last value in index=%d. Must provide positions in ascending order.",
-              logPosition, lastVirtualPosition);
+              blockPosition, lastVirtualPosition);
       throw new IllegalArgumentException(errorMessage);
     }
 
-    lastVirtualPosition = logPosition;
-
-    // write next entry
-    indexBuffer.putLong(entryLogPositionOffset(entryOffset), logPosition);
-    indexBuffer.putLong(entryAddressOffset(entryOffset), storageAddr);
+    lastVirtualPosition = blockPosition;
+    key.wrapLong(blockPosition);
+    value.wrapLong(blockAddress);
 
-    // increment size
-    indexBuffer.putIntOrdered(indexSizeOffset(), newIndexSize);
-
-    return newIndexSize;
+    blockPositionToAddress.put(key, value);
   }
 
-  private AtomicBuffer allocateBuffer(int capacity) {
-    final int requiredBufferCapacity = dataOffset() + (capacity * entryLength());
-    return bufferAllocator.apply(requiredBufferCapacity);
+  @Override
+  public void writeSnapshot(final String name, final long snapshotEventPosition) {
+    final String snapshotFileName =
+        snapshotConfiguration.snapshotFileName(name, snapshotEventPosition);
+    LOG.debug("Creating snapshot in {}", snapshotFileName);
+    db.createSnapshot(new File(snapshotFileName));
   }
 
-  private void expandIndexBuffer() {
-    final int newCapacity = Math.toIntExact(Math.round(capacity * GROW_FACTOR));
-    final AtomicBuffer newBuffer = allocateBuffer(newCapacity);
+  @Override
+  public void recoverFromSnapshot(final String name) throws Exception {
+    if (runtimeDirectory.exists()) {
+      FileUtil.deleteFolder(runtimeDirectory.getAbsolutePath());
+    }
 
-    newBuffer.putBytes(0, indexBuffer.byteArray());
+    final File snapshotDir = getLatestSnapshotDirectory(name, snapshotConfiguration);
```

Comment:
Seems really similar of the StateSnapshotController, maybe we can somehow reuse some code from there? Furthermore maybe we can use the fact that the snapshots contain the  snapshot position to find the latest snapshot.
This is also done in the StateSnapshotController:

```java

          snapshots.stream()
              .sorted(Comparator.reverseOrder())
              .filter(filter)
              .findFirst()
              .orElse(null);
```